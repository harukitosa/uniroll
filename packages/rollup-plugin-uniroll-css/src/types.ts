export type Options = {
  postprocess?: (css: string) => Promise<string>;
  /**
   * output css file instead of js injection.
   */
  output?: string;
};
