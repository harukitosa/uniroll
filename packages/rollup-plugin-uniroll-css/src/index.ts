import { Options } from "./types";
import { Plugin } from "rollup";
import { transformCss } from "./transformCss";
import { wrapWithStyleInjector } from "./wrapWithInjector";

export { wrapWithStyleInjector };

export const css = (options: Options = {}) => {
  const willOutputFile = !!options.output;
  let cssFiles: string[] = [];

  return {
    name: "uniroll-css",
    buildStart() {
      cssFiles = [];
    },
    async transform(code: string, id: string): Promise<string | void> {
      if (!id.endsWith(".css")) return;
      const css = await transformCss(code, options);

      if (willOutputFile) {
        cssFiles.push(css);
        return;
      } else {
        return wrapWithStyleInjector(css);
      }
    },
    generateBundle() {
      if (!willOutputFile) return;
      if (cssFiles.length === 0) return;
      this.emitFile({
        type: "asset",
        fileName: options.output,
        source: cssFiles.join(""),
      });
    },
  } as Plugin;
};
