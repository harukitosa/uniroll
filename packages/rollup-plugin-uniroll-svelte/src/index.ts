import { PreprocessorGroup } from "svelte/types/compiler/preprocess";
import type { EmittedAsset, Plugin } from "rollup";
import { compile as svelteCompile, preprocess } from "svelte/compiler";
import {
  cdnRewriteTransformerFactory,
  createSveltePreprocessor,
} from "./tsPreprocess";
import ts from "typescript";
import { CompileOptions } from "svelte/types/compiler/interfaces";
import { defaultResolveIdFallback } from "uniroll";
import { ResolveIdFallback } from "./types";

type SveltePluginOptions = {
  target?: ts.ScriptTarget;
  resolveIdFallback?: ResolveIdFallback;
  svelteOptions?: CompileOptions;
  extraPreprocess?: PreprocessorGroup[];
  emitCss?: boolean;
};

export const svelte = ({
  target = ts.ScriptTarget.ES2019,
  resolveIdFallback = defaultResolveIdFallback,
  extraPreprocess = [],
  emitCss = false,
  svelteOptions = {},
}: SveltePluginOptions = {}) => {
  return {
    name: "uniroll-svelte",
    async transform(code: string, id: string) {
      if (id.endsWith(".svelte")) {
        const tsPreprocess = createSveltePreprocessor({
          target,
          importer: id,
          resolveIdFallback: resolveIdFallback ?? defaultResolveIdFallback,
        });
        const { code: preprocessed } = await preprocess(
          code,
          [tsPreprocess, ...extraPreprocess],
          {
            filename: id,
          }
        );
        const result = svelteCompile(preprocessed, svelteOptions);
        result.warnings.forEach(warning => {
          const start = warning.start;
          if (start !== undefined) {
            const line = start.line;
            const column = start.column;
            this.warn(warning.message + " (" + line + ":" + column + ")");
          } else {
            this.warn(warning.message);
          }
        });

        if (emitCss && result.css.code) {
          const hash = Math.random().toString(32).substring(2);
          this.emitFile({
            type: "asset",
            source: result.css.code,
            fileName: `uniroll-svelte-${hash}.css`,
          } as EmittedAsset);
        }

        const tsCompiled = ts.transpileModule(result.js.code, {
          fileName: "$temp.tsx",
          compilerOptions: {
            target: target,
            module: ts.ModuleKind.ESNext,
          },
          transformers: {
            before: [cdnRewriteTransformerFactory(resolveIdFallback, id)],
          },
        });

        return {
          code: tsCompiled.outputText,
          map: tsCompiled.sourceMapText,
        };
      }
    },
  } as Plugin;
};
