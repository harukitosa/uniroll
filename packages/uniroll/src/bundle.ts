import { virtualFs } from "rollup-plugin-virtual-fs";
import { httpResolve, ResolveIdFallback } from "rollup-plugin-http-resolve";
import replace from "@rollup/plugin-replace";
import workerPlugin from "./plugins/worker-plugin";
import {
  defaultCache,
  defaultCompilerOptions,
  defaultDefine,
  defaultResolveIdFallback,
  transform,
} from "./shared";
import { rollup } from "rollup";
import json from "@rollup/plugin-json";
import { CompileOptions } from "./types";

export function bundle({
  input,
  files,
  cache = defaultCache,
  define = defaultDefine,
  resolveIdFallback = defaultResolveIdFallback,
  compilerOptions = {},
  extraPlugins = [],
  rollupOptions,
}: CompileOptions) {
  return rollup({
    input,
    plugins: [
      ...extraPlugins,
      json(),
      replace({ ...define }),
      workerPlugin(),
      httpResolve({
        cache,
        resolveIdFallback,
      }),
      virtualFs({ files }),
      transform({
        resolveIdFallback,
        compilerOptions: {
          ...defaultCompilerOptions,
          ...compilerOptions,
        },
      }),
    ],
    ...rollupOptions,
  });
}
