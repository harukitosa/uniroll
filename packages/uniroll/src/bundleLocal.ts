import { virtualFs } from "rollup-plugin-virtual-fs";
import replace from "@rollup/plugin-replace";
import workerPlugin from "./plugins/worker-plugin";
import {
  defaultCompilerOptions,
  defaultDefine,
  defaultExtractExternal,
  defaultResolveIdFallback,
  extractImportSpecfiers,
  transform,
} from "./shared";
import { rollup } from "rollup";
import json from "@rollup/plugin-json";
import { CompileOptions } from "./types";

export function bundleLocal({
  input,
  files,
  define = defaultDefine,
  compilerOptions = defaultCompilerOptions,
  extraPlugins = [],
  rollupOptions,
  resolveIdFallback = defaultResolveIdFallback,
  extractExternal = defaultExtractExternal,
}: CompileOptions & {
  extractExternal?: (specifiers: string[]) => string[];
}) {
  const spcefiers = extractImportSpecfiers(Object.values(files));
  const external = extractExternal(spcefiers);
  return rollup({
    input,
    external,
    plugins: [
      ...extraPlugins,
      json(),
      replace({ ...defaultDefine, ...define }),
      workerPlugin(),
      virtualFs({ files }),
      transform({
        resolveIdFallback,
        compilerOptions,
      }),
    ],
    ...rollupOptions,
  });
}
