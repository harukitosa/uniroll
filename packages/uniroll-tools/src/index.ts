import type { Plugin } from "rollup";

import { Command, flags } from "@oclif/command";
import path from "path";
import fs from "fs";
import { CompileOptions, getBundlePlugins } from "uniroll";
import { rollup, watch } from "rollup";
import fetch from "isomorphic-unfetch";
import { terser } from "rollup-plugin-terser";
import { extFallback } from "./extFallbackPlugin";
import { svelte } from "rollup-plugin-uniroll-svelte";
import ts from "typescript";

global.fetch = fetch;

const defaultConfig = {};

class UnirollCommand extends Command {
  static description = "uniroll cli builder";
  static flags = {
    version: flags.version({ char: "v" }),
    watch: flags.boolean({ char: "w" }),
    minify: flags.boolean({ char: "m" }),
    help: flags.help({ char: "h" }),
    output: flags.string({ char: "o", description: "output dir" }),
    format: flags.string({ char: "f", description: "output format" }),
    config: flags.string({ char: "c", description: "config path" }),
  };
  static args = [{ name: "input" }];

  async run() {
    const { args, flags } = this.parse(UnirollCommand);

    // load config
    const configPath = path.join(
      process.cwd(),
      flags.config ?? "uniroll.config.js"
    );
    let config = defaultConfig;
    if (fs.existsSync(configPath)) {
      config = require(configPath);
      this.log(`> Use config: ${configPath}`);
    }

    // input
    const inputPath = args.input.startsWith("/")
      ? args.input
      : path.join(process.cwd(), args.input);

    const { rollupOptions, input, ...others } = {
      ...config,
      input: inputPath,
      files: {}, // this is dummy,
      useVirtualFs: false,
      useNativeFs: true,
    } as CompileOptions;

    let plugins: Plugin[] = getBundlePlugins(others) as Plugin[];

    if (flags.minify) {
      plugins.push(terser({ module: true }) as Plugin);
    }
    plugins.push(extFallback({}));

    // include svelte default
    plugins.push(
      svelte({
        resolveIdFallback: others.resolveIdFallback,
        // target: ts.ScriptTarget.ES5,
      }) as Plugin
    );

    // output
    const format = (flags.format as any) ?? "es";
    let output = null;
    if (flags.output != null) {
      output = {
        format,
        dir: flags.output!.startsWith("/")
          ? flags.output
          : path.join(process.cwd(), flags.output),
      } as any;
    }
    if (flags.watch) {
      if (output == null) {
        this.error("uniroll --watch needs --output or --dir");
      }
      const watcher = watch({
        input,
        plugins,
        output,
        ...rollupOptions,
      });
      watcher.on("change", (path) => {
        console.log("[changed]", path);
      });
    } else {
      const rolled = await rollup({
        input,
        plugins,
      });
      if (output) {
        await rolled.write(output);
      } else {
        const out = await rolled.generate({
          format,
        });
        for (const output of out.output) {
          if (output.type === "asset") {
            console.log("//", output.fileName);
            console.log(output.source);
          } else {
            console.log("//", output.fileName);
            console.log(output.code);
          }
        }
      }
    }
  }
}

export = UnirollCommand;
