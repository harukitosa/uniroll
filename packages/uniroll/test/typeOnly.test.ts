import "isomorphic-unfetch";
import { bundle } from "../src/bundle";

test("build 1", async () => {
  const warned = [];
  const files = {
    "/foo.ts": "export type Foo = any;",
    "/index.tsx": "import type { Foo }  from './foo'; console.log(1 as Foo);",
  };

  const bundled = await bundle({
    files,
    input: "/index.tsx",
    rollupOptions: {
      onwarn(warnings, defaultHandler) {
        warned.push(warnings);
        defaultHandler(warnings);
      },
    },
  });
  expect(warned.length).toBe(0);
  const out = await bundled.generate({ format: "es" });
  console.log(out.output[0].code);
  expect(out.output[0]).toMatchSnapshot();
});

test("build 2", async () => {
  const warned = [];
  const files = {
    "/foo.ts": "export type Foo = any; export const x = 1;",
    // "/index.tsx": "import { Foo, x }  from './foo'; console.log(x as Foo);",
    "/index.tsx": "import { x } from './foo'; console.log(x);",
  };

  const bundled = await bundle({
    files,
    input: "/index.tsx",
    rollupOptions: {
      onwarn(warnings, defaultHandler) {
        warned.push(warnings);
        defaultHandler(warnings);
      },
    },
  });
  expect(warned.length).toBe(0);
  const out = await bundled.generate({ format: "es" });
  // console.log(out.output[0].code);
  expect(out.output[0]).toMatchSnapshot();
});
