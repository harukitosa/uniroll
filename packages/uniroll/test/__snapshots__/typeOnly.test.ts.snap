// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build 1 1`] = `
Object {
  "code": "console.log(1);
",
  "dynamicImports": Array [],
  "exports": Array [],
  "facadeModuleId": "/index.tsx",
  "fileName": "index.js",
  "implicitlyLoadedBefore": Array [],
  "importedBindings": Object {},
  "imports": Array [],
  "isDynamicEntry": false,
  "isEntry": true,
  "isImplicitEntry": false,
  "map": null,
  "modules": Object {
    "/index.tsx": Object {
      "originalLength": 57,
      "removedExports": Array [],
      "renderedExports": Array [],
      "renderedLength": 15,
    },
  },
  "name": "index",
  "referencedFiles": Array [],
  "type": "chunk",
}
`;

exports[`build 2 1`] = `
Object {
  "code": "const x = 1;

console.log(x);
",
  "dynamicImports": Array [],
  "exports": Array [],
  "facadeModuleId": "/index.tsx",
  "fileName": "index.js",
  "implicitlyLoadedBefore": Array [],
  "importedBindings": Object {},
  "imports": Array [],
  "isDynamicEntry": false,
  "isEntry": true,
  "isImplicitEntry": false,
  "map": null,
  "modules": Object {
    "/foo.ts": Object {
      "originalLength": 42,
      "removedExports": Array [],
      "renderedExports": Array [
        "x",
      ],
      "renderedLength": 12,
    },
    "/index.tsx": Object {
      "originalLength": 42,
      "removedExports": Array [],
      "renderedExports": Array [],
      "renderedLength": 15,
    },
  },
  "name": "index",
  "referencedFiles": Array [],
  "type": "chunk",
}
`;
